app/scripts/migrations/081.ts

```diff
@@ -0,0 +1,162 @@
+import { cloneDeep, isArray } from 'lodash';
+import {
+  WALLET_SNAP_PERMISSION_KEY,
+  SnapCaveatType,
+} from '@metamask/rpc-methods';
+import { hasProperty, isObject } from '@metamask/utils';
+
+export const version = 81;
+
+/**
+ * Prior to this migration, snap <> dapp permissions were wildcards i.e. `wallet_snap_*`.
+ * Now the permission has been changed to `wallet_snap` and the current snap permissions
+ * that are under wildcards will be added as caveats to a parent `wallet_snap` permission.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(originalVersionedData: {
+  meta: { version: number };
+  data: Record<string, unknown>;
+}) {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  const state = versionedData.data;
+  const newState = transformState(state);
+  versionedData.data = newState;
+  return versionedData;
+}
+
+// We return state AS IS if there is any corruption
+function transformState(state: Record<string, unknown>) {
+  if (
+    !hasProperty(state, 'PermissionController') ||
+    !isObject(state.PermissionController)
+  ) {
+    return state;
+  }
+  const { PermissionController } = state;
+
+  const { subjects } = PermissionController;
+
+  if (!isObject(subjects)) {
+    return state;
+  }
+
+  const snapPrefix = 'wallet_snap_';
+
+  for (const [subjectName, subject] of Object.entries(subjects)) {
+    if (!isObject(subject) || !isObject(subject.permissions)) {
+      return state;
+    }
+    // We keep track of the latest permission's date and associated id
+    // to assign to the wallet_snap permission after iterating through all permissions
+    let date = 1;
+    let id;
+    const { permissions } = subject;
+    // New permissions object that we use to tack on the `wallet_snap` permission
+    const updatedPermissions = { ...permissions };
+    for (const [permissionName, permission] of Object.entries(permissions)) {
+      // check if the permission is namespaced
+      if (permissionName.startsWith(snapPrefix)) {
+        if (
+          !isObject(permission) ||
+          !hasProperty(permission, 'id') ||
+          !hasProperty(permission, 'date')
+        ) {
+          return state;
```

Comment:
Is the idea that if we see anything out of the ordinary we return the state and hopefully nothing is broken?

Reply:
Yeah, basically doing what Mark said: https://github.com/MetaMask/metamask-extension/pull/17718#discussion_r1122131559